[WIP] Adds BTF CO:RE functionality to KubeArmor

AllMight2099 opened the pull request:
Fixes #789 
Signed-off-by: Nishanth R <nishusheela2011@gmail.com>

AllMight2099 committed:
Uploads system monitor binary and embeds it into the kubearmor binary

Signed-off-by: Nishanth R <nishusheela2011@gmail.com>

daemon1024 commented on KubeArmor/monitor/systemMonitor.go:
```diff
@@ -118,6 +120,9 @@ type ContextCombined struct {
 // StopChan Channel
 var StopChan chan struct{}
 
+//go:embed system_monitor.*
```
We should nams these binaries different so they are differentiated from runtime compiled binaries in general

AllMight2099 replied:
I've changed the binary names to `embedded_system_monitor` to reflect that this would be embedded inside the Kubearmor binary

daemon1024 commented on KubeArmor/monitor/systemMonitor.go:
```diff
@@ -251,22 +256,44 @@ func (mon *SystemMonitor) InitBPF() error {
+	var bpfFile io.ReaderAt
 
+	mon.Logger.Print("Initializing eBPF system monitor")
+	bpfPath = os.Getenv("PWD") + "/monitor/"
```
This is breaking, why is there a need to change bpfPath

AllMight2099 replied:
Oh shoot, I forgot to revert that, I was using that to test if I could use a pre-compiled system_monitor binary that I saved in a different location. I'll revert it

daemon1024 commented on KubeArmor/monitor/systemMonitor.go:
```diff
@@ -251,22 +256,44 @@ func (mon *SystemMonitor) InitBPF() error {
 		bpfPath = bpfPath + "system_monitor.bpf.o"
 	}
 	mon.Logger.Printf("eBPF system monitor object file path: %s", bpfPath)
-	mon.BpfModule, err = cle.LoadCollection(bpfPath)
```
We need to keep this for fall back?

AllMight2099 replied:
I've made these changes

daemon1024 commented on KubeArmor/monitor/systemMonitor.go:
```diff
@@ -251,22 +256,44 @@ func (mon *SystemMonitor) InitBPF() error {
+		return fmt.Errorf("unable to load collection spec from %v", err)
+	}
+
+	mon.BpfModule, err = cle.NewCollection(spec)
```
We should load this spec only if BTF is available, we need to add a check for BTF information check before load embed binaries else need to look at runtime config.

AllMight2099 replied:
I've made these changes too

daemon1024 reviewed (changes requested):
Thanks for the PR! Find review comments inline.

Noting existing functionality shouldn't be breaking ( it is right now as suggested by the failing tests ). This should be an incremental change.

AllMight2099 committed:
Adds fallback condition to check if btf information is available

Signed-off-by: Nishanth R <nishusheela2011@gmail.com>

AllMight2099 committed:
Updates name of the system_monitor binaries

Signed-off-by: Nishanth R <nishusheela2011@gmail.com>

AllMight2099 committed:
1. Updates embedded_system_monitor binaries
2. Removes go clean -testcache when testing system monitor
3. fixes btf file check in systemMonitor.go

Signed-off-by: Nishanth R <nishusheela2011@gmail.com>

AllMight2099 committed:
Adds script that downloads and generated reduced btf files from btfhub

Signed-off-by: Nishanth R <nishusheela2011@gmail.com>

AllMight2099 committed:
Fixes if condition

Signed-off-by: Nishanth R <nishusheela2011@gmail.com>

AllMight2099 committed:
Adds code to get OS Info to get reduced btf file

Signed-off-by: Nishanth R <nishusheela2011@gmail.com>

AllMight2099 committed:
adds reduced btf-files

Signed-off-by: Nishanth R <nishusheela2011@gmail.com>

AllMight2099 committed:
minor fix

Signed-off-by: Nishanth R <nishusheela2011@gmail.com>

AllMight2099 committed:
minor fix

Signed-off-by: Nishanth R <nishusheela2011@gmail.com>

AllMight2099 committed:
Adds OSInfo and removes reduced btf files

Signed-off-by: Nishanth R <nishusheela2011@gmail.com>

daemon1024 commented on btf-utils/btfhub-btfgen.sh:
```diff
@@ -0,0 +1,60 @@
+#!/bin/bash -e
```
Where is this script called?

AllMight2099 replied:
It's not called anywhere, I made it to download btfhub-archive and generate the reduced files

daemon1024 replied:
How would we ship it with our container? It should be called at docker build time tho right?

AllMight2099 replied:
This was meant to run everytime someone changed system_monitor.c file. I was going to store the reduced files as a zip and unzip it in the container

daemon1024 replied:
Wouldn't generating it as part of container image cover the case for 
> run everytime someone changed system_monitor.c file

We can assume that no one would expect BTF files when using non CORE in development environment.

AllMight2099 replied:
Generating it in the container means having btfhub-archive, which isn't feasible cause it takes around 5-ish GB

daemon1024 replied:
Hey @AllMight2099 , We have a misunderstanding here.

I am saying it to generate it at build time, not runtime. Like how we compile the KubeArmor code to a single binary in our Dockerfile and ship only the binary.

Similarly we download and reduce the btfhub at build time and copy the reduced files from staging to main container using Dockerfile.

AllMight2099 replied:
Ohhh, got it

daemon1024 replied:
For reference

https://github.com/kubearmor/KubeArmor/blob/fc08819ea1efc8d3c68b399d7b73fe44859baab6/Dockerfile#L18 built in builder image

https://github.com/kubearmor/KubeArmor/blob/fc08819ea1efc8d3c68b399d7b73fe44859baab6/Dockerfile#L32
Copied from builder image to actual image.

AllMight2099 replied:
I'll make those changes

AllMight2099 committed:
Completes btfhub-btfgen script

Signed-off-by: Nishanth R <nishusheela2011@gmail.com>

AllMight2099 committed:
Adds btfhub files to dockerignore

Signed-off-by: Nishanth R <nishusheela2011@gmail.com>
